crates/bevy_asset/src/io/memory.rs

```diff
@@ -361,6 +409,183 @@ impl AssetReader for MemoryAssetReader {
     }
 }
 
+/// A writer that writes into [`Dir`], buffering internally until flushed/closed.
+struct DataWriter {
+    /// The dir to write to.
+    dir: Dir,
+    /// The path to write to.
+    path: PathBuf,
+    /// The current buffer of data.
+    ///
+    /// This will include data that has been flushed already.
+    current_data: Vec<u8>,
+    /// Whether to write to the data or to the meta.
+    write_meta: bool,
+}
+
+impl AsyncWrite for DataWriter {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        _: &mut core::task::Context<'_>,
+        buf: &[u8],
+    ) -> Poll<std::io::Result<usize>> {
+        self.get_mut().current_data.extend_from_slice(buf);
+        Poll::Ready(Ok(buf.len()))
+    }
+
+    fn poll_flush(
+        self: Pin<&mut Self>,
+        _: &mut core::task::Context<'_>,
+    ) -> Poll<std::io::Result<()>> {
+        // Write the data to our fake disk. This means we will repeatedly reinsert the asset.
```

Comment:
Im not sure I understand why there's both a DataWriter and MemoryAssetWriter. Is DataWriter also a memory writer? If not, why is this a fake disk?

Reply:
MemoryAssetWriter implements the AssetWriter trait - which is really an interface over the file system to start writes, move directories, etc.

DataWriter implements the AsyncWrite trait that actually writes data. 

This just mirrors the existing MemoryAssetReader and DataReader types. I'd like to keep that consistent.